tokenawarebandit.go

```diff
@@ -0,0 +1,614 @@
+	// getKeyspaceMetadata is used to mock Session.KeyspaceMetadata in tests.
+	getKeyspaceMetadata func(keyspace string) (*KeyspaceMetadata, error)
+	// afterReplicasUpdates is called after replicas are updated, used only for tests.
+	afterReplicasUpdated func(keyspace string)
+
+	logger StdLogger
+	// explorationPortion is a portion of queries (number 0..1) to use for exploration.
+	// We need some portion of queries to determine the latency of other nodes.
+	explorationPortion float32
+	// decayPeriod is duration over which the sampled latency values linearly decay.
+	decayPeriod time.Duration
+	// localDatacenter is name of the local datacenter.
+	localDatacenter string
+	// remoteDatacenterPenalty is added to the latency of nodes in remote datacenters.
+	remoteDatacenterPenalty time.Duration
+
+	// stopChan is closed when Stop is called.
+	stopChan chan struct{}
+	// stopOnce is used to close stopChan only once.
+	stopOnce sync.Once
+
+	// mu protects fields below.
+	// If you need to hold both TokenAwareBanditHostPolicy.mu and tokenAwareBanditHost.mu,
+	// TokenAwareBanditHostPolicy.mu must be locked before tokenAwareBanditHost.mu.
+	mu sync.RWMutex
+
+	// hosts by HostID().
+	// contents of tokenAwareBanditHost are not protected by mu as it has its own lock.
+	hosts map[string]*tokenAwareBanditHost
+	// keyspaces by name.
+	keyspaces map[string]*tokenAwareBanditKeyspace
+	// tokenRing to use for mapping tokens to hosts.
+	tokenRing *tokenRing
+	// partitioner used to map partition keys to tokens.
+	partitioner string
+}
+
+type tokenAwareBanditKeyspace struct {
+	// replicas allows to get replicas for token.
+	replicas tokenRingReplicas
+
+	// updating is true if we are currently updating the replica map for this keyspace.
+	updating bool
+
+	// needsUpdate is true if something changed while updating the replica map.
+	needsUpdate bool
+}
+
+// tokenAwareBanditHost keeps track of per-host state.
+type tokenAwareBanditHost struct {
+	// mu protects fields in tokenAwareBanditHost.
+	// If you need to hold both TokenAwareBanditHostPolicy.mu and tokenAwareBanditHost.mu,
+	// TokenAwareBanditHostPolicy.mu must be locked before tokenAwareBanditHost.mu.
+	mu sync.RWMutex
+	// latencyBuckets is a cyclic buffer storing history of latency measurements for a host.
+	// latencyBuckets[currentBucket] is the newest entry, the one to the right (with wrap) is the oldest entry.
+	latencyBuckets [bucketCount]latencyBucket
+	// currentBucket points to the newest bucket (the one we are writing to).
+	currentBucket int
+	// hostInfo of the host.
+	hostInfo *HostInfo
+}
+
+type latencyBucket struct {
+	sum   time.Duration
+	count int64
+}
+
+func (lb *latencyBucket) add(latency time.Duration) {
+	lb.sum += latency
+	lb.count += 1
+}
+
+// latency returns weighted average of latency measurements over time.
+// weight is linear from oldest bucket (1/bucketCount) to newest bucket (1.0).
+// h.mu must be held at least read-only.
```

Comment:
typo probably
```suggestion
// th.mu must be held at least read-only.
```

could we plz also mention when the `ok` is false and what does it mean?